### ISSUE TITLE
Add ApplyBuilder syntax

### ISSUE THREAD

**Issue by mpilquist** [non/cats#162]
Add infix synatx for calling `applyN`, `mapN`, and `tupledN` on an `Apply[F]`.

Something like:
```scala
trait ApplyBuilder2[F[_], A, B] {
  val a: F[A]
  val b: F[B]
  def map[Z](f: (A, B) => C)(implicit F: Apply[F]): F[Z] = F.map2(a, b)(f)
  def apply[Z](f: F[(A, B) => C])(implicit F: Apply[F]): F[Z] = F.apply2(a, b)(f)
  def tupled(implicit F: Apply[F]): F[(A, B)] = F.tupled(a, b)
  def |@|(c0: F[C]): ApplyBuilder3[F, A, B, C] = new ApplyBuilder3[C] {
    val c: F[C] = c0
  }

  trait ApplyBuilder3[C] {
    val c: F[C]
    ...
  }
}
```

This differs a little from the Scalaz version in that `apply` maps to `applyN` and `map` maps to `mapN`. As a reminder, cats' `mapN` is the same as Scalaz's `applyN` and cats' `applyN` is the same as Scalaz's `apN`.

**Comment by stephenjudkins** [comment]
This is really ugly in Scalaz. Any thoughts on using quasiquotes for this kind of boilerplate? Is it worth it?

Here's a quick example showing that it might be possible to use macros to make boilerplate-heavy stuff like this readable: https://gist.github.com/stephenjudkins/796aacbe798f54269890

**Comment by stew** [comment]
my only ask is that we keep good type inference as a high priority

**Comment by stephenjudkins** [comment]
I'll give it a shot. If people are appalled we can go the copypasta route.

**Comment by tpolecat** [comment]
@stew code of conduct forbids using "ask" as a noun. cut it out.

**Comment by stephenjudkins** [comment]
I'm hoping my approach will give us some win

**Comment by stephenjudkins** [comment]
Is the consensus here that @travisbrown 's approach is the way to go? Easily gives us 2.10.x support, and no more complicated than the macro.

**Comment by ceedubs** [comment]
Fixed by #176
